lumicks/pylake/detail/widefield.py

```diff
@@ -327,3 +299,121 @@ def crop(self, roi):
     @property
     def shape(self):
         return self.y_max - self.y_min, self.x_max - self.x_min
+
+
+@dataclass
+class TransformMatrix:
+    """Wrapper around affine transformation matrix with convenience functions."""
+
+    matrix: np.ndarray = field(default=np.eye(3)[:2, :])
+
+    def __mul__(self, mat):
+        """Perform matrix multiplication such that `self * mat == np.matmul(self, mat)`."""
+        assert isinstance(mat, TransformMatrix), "Operands must be of type `TransformMatrix`."
+        m1 = np.vstack((self.matrix, [0, 0, 1]))
+        m2 = np.vstack((mat.matrix, [0, 0, 1]))
+        return TransformMatrix(np.matmul(m1, m2)[:2])
+
+    @classmethod
+    def with_offset(cls, alignment, x_offset, y_offset):
+        """Recalculate matrix with offset. Same implementation as Bluelake.
+
+        Parameters
+        ----------
+        alignment: np.ndarray
+            alignment matrix read from Bluelake metadata
+        x_offset: float
+            alignment ROI x offset
+        y_offset: float
+            alignment ROI y offset
+        """
+        if x_offset == 0 and y_offset == 0:
+            return cls(alignment)
+
+        # translate the origin of the image so that it matches that of the original transform
+        translation = np.eye(3)
+        translation[0, -1] = -x_offset
+        translation[1, -1] = -y_offset
+        # apply the original transform to the translated image.
+        # it only needs to be resized from a 2x3 to a 3x3 matrix
+        original = np.vstack((alignment, [0, 0, 1]))
+        # translate the image back to the original origin.
+        # takes into account both the offset and the scaling performed by the first step
+        back_translation = np.eye(3)
+        back_translation[0, -1] = original[0, 0] * x_offset
+        back_translation[1, -1] = original[1, 1] * y_offset
+        # concatenate the transforms by multiplying their matrices and ignore the unused 3rd row
+        return cls(np.dot(back_translation, np.dot(original, translation))[:2, :])
+
+    @classmethod
+    def rotation(cls, theta, center):
+        """Construct matrix for rotation by angle `theta` about a point `center`.
+
+        Parameters
+        ----------
+        theta: float
+            angle of rotation
+        center: np.ndarray
+            (x, y) point, center of rotation
+        """
+        # form individual translation and rotation matrices
+        translate_to_center = np.eye(3)
+        translate_to_center[:2, -1] = -np.array(center)
+
+        rotate = np.eye(3)
+        rotate[:2, :2] = [[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]]
+
+        translate_to_origin = np.eye(3)
+        translate_to_origin[:2, -1] = center
+
+        # multiply to get overall transformation matrix
+        return cls(np.matmul(translate_to_origin, np.matmul(rotate, translate_to_center))[:2, :])
+
+    def warp_image(self, data, inverse=True, interp_mode=cv2.INTER_LINEAR):
```

Comment:
👍 added an `invert()` method and got rid of the `interp_mode` altogether since we do a single linear interpolation now to take care of both color channel alignment and rotation about the tether